src/lib/format/l20n/parser.js

```diff
@@ -5,201 +5,126 @@ import { L10nError } from '../../errors';
 const MAX_PLACEABLES = 100;
 
 export default {
-  _patterns: null,
-
-  init: function() {
-    this._patterns = {
-      index: /@cldr\.plural\(\$?(\w+)\)/g,
-      placeables: /\{\{\s*\$?([^\s]*?)\s*\}\}/,
-      unesc: /\\({{|u[0-9a-fA-F]{4}|.)/g,
-    };
-  },
-
-  parse: function (emit, string, simple) {
-    if (!this._patterns) {
-      this.init();
-    }
+  parse: function(env, string) {
     this._source = string;
     this._index = 0;
-    this._length = this._source.length;
-    this.simpleMode = simple;
-    this.emit = emit;
+    this._length = string.length;
+    this.entries = Object.create(null);
 
-    return this.getL20n();
+    return this.getResource();
```

Comment:
I'm willing to give your approach a try, although I suspect it will cost more.
But I'd prefer not to mix this refactor into this patch. I'll take it as a separate one after this, ok?